### PR TITLE
Bmft angular8 upgrade

### PR DESCRIPTION
**Pull request by MoeMaher**

### CODE REVIEW

File: Frontend/src/app/app.component.html [hunk 1]
```diff
@@ -14,9 +14,9 @@ <h1>Bing Maps Fleet Tracker</h1>
 </section>
 
 <mat-sidenav-container>
-  <mat-sidenav #sidenav mode="side" opened="true" disableClose="true" [appRoleLimiter]="Roles.Viewer">
+  <mat-sidenav #sidenav mode="side" opened="true" disableClose="true" [appRoleLimiter]="Roles.Viewer" style="overflow-x: hidden!important;">
```
Comment (MustafaLeithy) [inline Frontend/src/app/app.component.html]: Avoid using "style" and use css classes for better code maintainability.

File: Frontend/src/app/assets/asset.service.spec.ts [hunk 2]
```diff
@@ -4,14 +4,17 @@
 /* tslint:disable:no-unused-variable */
 
 import { TestBed, inject } from '@angular/core/testing';
+// import { Observable } from 'rxjs';
```
Comment (momohs) [inline Frontend/src/app/assets/asset.service.spec.ts]: not needed?

File: Frontend/src/app/assets/asset.service.ts [hunk 3]
```diff
@@ -1,16 +1,22 @@
+
+import {map} from 'rxjs/operators';
```
Comment (momohs) [inline Frontend/src/app/assets/asset.service.ts]: imports should go beneath header

File: Frontend/src/app/assets/asset.service.ts [hunk 4]
```diff
@@ -38,14 +44,14 @@ export class AssetService {
   }
 
   getPoints(id: string, dateRange?: DateRange): Observable<TrackingPoint[]> {
-    return this.dataService.get<TrackingPoint>(`assets/${id}/points`)
-      .map(points => {
+    return this.dataService.get<TrackingPoint>(`assets/${id}/points`).pipe(
```
Comment (momohs) [inline Frontend/src/app/assets/asset.service.ts]: why mix and match between rxjs 6 and rxjs5 syntax?

File: Frontend/src/app/assets/asset.service.ts [hunk 5]
```diff
@@ -58,15 +64,15 @@ export class AssetService {
   }
 
   getTrips(id: string, dateRange?: DateRange): Observable<Trip[]> {
-    return this.dataService.get<Trip>(`assets/${id}/trips`)
-      .map(trips => {
+    return this.dataService.get<Trip>(`assets/${id}/trips`).pipe(
```
Comment (momohs) [inline Frontend/src/app/assets/asset.service.ts]: ditto. Unify the syntax

File: Frontend/src/app/core/data.service.ts [hunk 6]
```diff
@@ -1,16 +1,16 @@
+
+import {map} from 'rxjs/operators';
```
Comment (momohs) [inline Frontend/src/app/core/data.service.ts]: imports after header

File: Frontend/src/app/core/data.service.ts [hunk 7]
```diff
@@ -29,8 +29,8 @@ export class DataService {
         const cache = this.getCache<T>(path);
         const url = this.getUrl(path);
 
-        this.authHttpService.get(url)
-            .map(response => response.json() as T[])
+        this.authHttpService.get(url).pipe(
+            map(response => response as unknown as T[]))
```
Comment (momohs) [inline Frontend/src/app/core/data.service.ts]: Why is this needed? (repeated across file)

File: Frontend/src/app/core/settings.service.ts [hunk 8]
```diff
@@ -1,9 +1,11 @@
+
+import {map} from 'rxjs/operators';
```
Comment (momohs) [inline Frontend/src/app/core/settings.service.ts]: ditto

File: Frontend/src/app/devices/device-editor/device-editor.component.ts [hunk 9]
```diff
@@ -53,7 +58,8 @@ export class DeviceEditorComponent implements OnInit, OnDestroy {
       if (id) {
         this.isEditing = true;
         this.deviceSubscription = this.deviceService.getDevice(id).subscribe(device => this.device = device);
-        this.deviceService.getToken(id).take(1).subscribe(t => this.deviceToken = t);
+        var obs: any = this.deviceService.getToken(id)
+        obs.take(1).subscribe(t => this.deviceToken = t);
```
Comment (momohs) [inline Frontend/src/app/devices/device-editor/device-editor.component.ts]: keep as single line.

File: Frontend/src/app/devices/device-editor/device-editor.component.ts [hunk 10]
```diff
@@ -71,7 +77,10 @@ export class DeviceEditorComponent implements OnInit, OnDestroy {
         }
       });
 
-    this.hubConnection = new HubConnection(this.deviceService.getDeviceAdditionNotificationUrl());
+    // this.hubConnection = new HubConnection(this.deviceService.getDeviceAdditionNotificationUrl());
+    this.hubConnection =  new signalR.HubConnectionBuilder()
+    .withUrl(this.deviceService.getDeviceAdditionNotificationUrl())
+    .build();
```
Comment (momohs) [inline Frontend/src/app/devices/device-editor/device-editor.component.ts]: indentation errors

File: Frontend/src/app/devices/device-editor/device-editor.component.ts [hunk 11]
```diff
@@ -71,7 +77,10 @@ export class DeviceEditorComponent implements OnInit, OnDestroy {
         }
       });
 
-    this.hubConnection = new HubConnection(this.deviceService.getDeviceAdditionNotificationUrl());
+    // this.hubConnection = new HubConnection(this.deviceService.getDeviceAdditionNotificationUrl());
```
Comment (momohs) [inline Frontend/src/app/devices/device-editor/device-editor.component.ts]: not needed

File: Frontend/src/app/devices/device.service.ts [hunk 12]
```diff
@@ -1,8 +1,10 @@
+
+import {map} from 'rxjs/operators';
```
Comment (momohs) [inline Frontend/src/app/devices/device.service.ts]: ditto

File: Frontend/src/app/devices/device.service.ts [hunk 13]
```diff
@@ -34,14 +36,14 @@ export class DeviceService {
   }
 
   getPoints(id: string, dateRange?: DateRange): Observable<TrackingPoint[]> {
-    return this.dataService.get<TrackingPoint>(`devices/${id}/points`)
-      .map(points => {
+    return this.dataService.get<TrackingPoint>(`devices/${id}/points`).pipe(
```
Comment (momohs) [inline Frontend/src/app/devices/device.service.ts]: ditto

File: Frontend/src/app/geofences/geofence-editor/geofence-editor.component.ts [hunk 14]
```diff
@@ -149,15 +149,19 @@ export class GeofenceEditorComponent implements OnInit, OnDestroy {
     });
 
     if (this.isEditable) {
+      console.log('inside if consition updating ')
```
Comment (momohs) [inline Frontend/src/app/geofences/geofence-editor/geofence-editor.component.ts]: testing code

File: Frontend/src/app/geofences/geofence-editor/geofence-editor.component.ts [hunk 15]
```diff
@@ -149,15 +149,19 @@ export class GeofenceEditorComponent implements OnInit, OnDestroy {
...
       this.geofenceService.update(this.geofence)
         .subscribe(() => {
           this.router.navigate(['/geofences']);
         });
     } else {
+      console.log('inside if consition adding ')
```
Comment (momohs) [inline Frontend/src/app/geofences/geofence-editor/geofence-editor.component.ts]: testing code

File: Frontend/src/app/geofences/geofence-editor/geofence-editor.component.ts [hunk 16]
```diff
@@ -149,15 +149,19 @@ export class GeofenceEditorComponent implements OnInit, OnDestroy {
...
       this.geofenceService.add(this.geofence)
         .subscribe(() => {
-          this.router.navigate(['/geofences']);
+          console.log('done adding')
```
Comment (momohs) [inline Frontend/src/app/geofences/geofence-editor/geofence-editor.component.ts]: testing code

File: Frontend/src/app/locations/location-editor/location-editor.component.ts [hunk 17]
```diff
@@ -80,7 +80,8 @@ export class LocationEditorComponent implements OnInit, OnDestroy {
   changeAddress(): void {
     const address = this.location.address;
     this.mapService.geocodeQuery(address);
-    this.mapService.getGeocodeResult().take(1).subscribe(point => {
+    var obs: any = this.mapService.getGeocodeResult()
+    obs.take(1).subscribe(point => {
```
Comment (momohs) [inline Frontend/src/app/locations/location-editor/location-editor.component.ts]: keep in single line

File: Frontend/src/app/maps/maps/maps.component.ts [hunk 18]
```diff
@@ -19,7 +19,7 @@ import { TrackingPoint } from '../../shared/tracking-point';
     styleUrls: ['./maps.component.css']
 })
 export class MapsComponent implements OnInit {
-    @ViewChild('map') mapElement: ElementRef;
+    @ViewChild('map',{static: false}) mapElement: ElementRef;
```
Comment (momohs) [inline Frontend/src/app/maps/maps/maps.component.ts]: why is this needed?

File: Frontend/src/app/users/login/login.component.ts [hunk 19]
```diff
@@ -70,18 +69,18 @@ export class LoginComponent implements OnInit, OnDestroy {
...
-                  width: '70%',
-                }).afterClosed().subscribe(res => {
-                  if (res != null) {
-                    this.settingsService.setInstrumentationApproval(res);
-                  }
-                });
-              }
-            });
+          var obs : any = this.settingsService.getInstrumentationApproval()
+          obs.take(1).subscribe(val => {
```
Comment (momohs) [inline Frontend/src/app/users/login/login.component.ts]: keep as a single line. i.e. no need to use `var obs`

File: Frontend/src/environments/environment.ts [hunk 20]
```diff
@@ -8,7 +8,7 @@
 
 export const environment = {
   production: false,
-  backendUrl: 'http://localhost:4488/api',
+  backendUrl: 'https://localhost:44396/api',
```
Comment (momohs) [inline Frontend/src/environments/environment.ts]: why change this?

File: Frontend/src/app/app.component.css [hunk 21]
```diff
@@ -50,4 +50,12 @@ h1 {
 .spinner {
   vertical-align: text-bottom;
   width: 25px;
-}
\ No newline at end of file
+}
+
+.mat-raised-button{
```
[the remark this answers is not in the document]
Comment (momohs) [reply Frontend/src/app/app.component.css]: I'd say it shouldn't cause a problem as long as it is inside the component css and not in the styles.css

[the remark this answers is not in the document]
Comment (MoeMaher) [reply Frontend/src/app/app.component.css]: there are small differences in the default styles of Angular material components after Upgrading the @angular/material dependency. All buttons in the application to reset the old default values.

is this acceptable, or should I find another workaround?

File: Frontend/src/app/assets/asset-editor/asset-editor.component.ts [hunk 22]
```diff
@@ -3,7 +3,7 @@
 
 import { Component, OnDestroy, OnInit } from '@angular/core';
 import { ActivatedRoute, Router } from '@angular/router';
-import { Subscription } from 'rxjs/Subscription';
+import { Subscription } from 'rxjs';
```
[the remark this answers is not in the document]
Comment (MoeMaher) [reply Frontend/src/app/assets/asset-editor/asset-editor.component.ts]: in Angular 6 yes

File: Frontend/src/app/assets/asset-editor/asset-editor.component.ts [hunk 23]
```diff
@@ -12,6 +12,7 @@ import { AssetProperties } from '../../shared/asset-properties';
 import { AssetService } from '../asset.service';
 import { Device } from '../../devices/device';
 import { DeviceService } from '../../devices/device.service';
+import 'rxjs/add/operator/takeWhile';
```
[the remark this answers is not in the document]
Comment (MoeMaher) [reply Frontend/src/app/assets/asset-editor/asset-editor.component.ts]: After the upgrade, we need to import takeWhile, skipWhile and all the rxjs/add/operators when needed

File: Frontend/src/app/assets/asset.service.spec.ts (hunk 2 above)
Comment (MoeMaher) [reply Frontend/src/app/assets/asset.service.spec.ts]: I fixed it thanks. "import { Subscription } from 'rxjs';" is used in angular 6+

File: Frontend/src/app/assets/asset.service.ts (hunk 3 above)
Comment (MoeMaher) [reply Frontend/src/app/assets/asset.service.ts]: I fixed it, Thanks

File: Frontend/src/app/maps/maps/maps.component.ts (hunk 18 above)
Comment (MoeMaher) [reply Frontend/src/app/maps/maps/maps.component.ts]: In Angular's post, in version 8, we are migrating all applications and libraries to explicitly specify the resolution strategy for @ViewChild and @ContentChild queries.

File: Frontend/src/app/core/data.service.ts (hunk 7 above)
Comment (MoeMaher) [reply Frontend/src/app/core/data.service.ts]: - In HttpClient, JSON is an assumed default and no longer needs to be explicitly parsed
- According to response, it is an Array of Objects but missing some properties from the normal array in javascript, like pop, push and concat. so I force typecast it discarding the remaining properties

Comment (momohs) [reply Frontend/src/app/core/data.service.ts]: I meant use get <T[]>() instead

File: Frontend/src/app/core/data.service.ts [hunk 24]
```diff
@@ -1,16 +1,17 @@
+
+
```
Comment (momohs) [inline Frontend/src/app/core/data.service.ts]: remove extra white space at the top of the file

File: Frontend/src/app/devices/device-editor/device-editor.component.ts [hunk 25]
```diff
@@ -3,10 +3,11 @@
 
 import { Component, OnDestroy, OnInit } from '@angular/core';
 import { ActivatedRoute, Router } from '@angular/router';
-import { MatDialog, MatDialogRef } from '@angular/material';
-import { Subscription } from 'rxjs/Subscription';
+import { MatDialog, MatDialogRef } from '@angular/material/dialog';
+import { Subscription } from 'rxjs';
 import { ToasterService } from 'angular2-toaster';
-import { HubConnection } from '@aspnet/signalr-client';
+// import { HubConnection } from '@aspnet/signalr-client';
```
Comment (momohs) [inline Frontend/src/app/devices/device-editor/device-editor.component.ts]: not needed?

File: Frontend/src/app/devices/device.service.ts (hunk 12 above)
Comment (momohs) [reply Frontend/src/app/devices/device.service.ts]: Move import below header

File: Frontend/src/app/app.component.css (hunk 21 above)
[the remark this answers is not in the document]
Comment (momohs) [reply Frontend/src/app/app.component.css]: I think it's ok
